Optimize multicast routing memory usage

kderda opened the pull request:
This PR adds new configurable functionality of adding multiple network interfaces to a single multicast routing entry. I tried to keep changes backward compatible by keeping `struct net_if *iface` in union. I have also implemented a new test case and changes in shell module.

An example use case: Tunneling packets between 5 interfaces for a given prefix previously required usage of 5 entries with the same prefix information.

kderda committed:
net: ipv6: check if multicast packet was forwarded

This commit adds a check to determine if the packet wasn't already
forwarded to a given interface.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

kderda committed:
net: ipv6: add multiple interfaces to multicast routing entry

In order to save memory, a single multicast routing entry now
contains configurable number of network interfaces.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

kderda committed:
tests: net: mcast_routing: add verification of multiple interfaces

This commit adds a test for veryfing new functionality of storing
mutliple interfaces per single multicast routing entry.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

kderda committed:
net: shell: print interfaces per multicast route entry

This commit change a method of printing multicast routes by showing
all interfaces per entry instead of aggregating them by interface.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

rlubos commented on subsys/net/ip/Kconfig:
```diff
@@ -426,6 +426,14 @@ config NET_MAX_MCAST_ROUTES
 
+config NET_MCAST_ROUTE_MAX_IFACES
+	int "Max number of network interfaces per multicast routing entry"
+	default 1
```
Could add a `range 1 xxx` here, so that it's not possible to misconfigure.

konradderda replied:
As far as I know, there is no way to fetch the number of network interfaces so I have set max to arbitrary chosen 8.

rlubos commented on subsys/net/ip/route.h:
```diff
@@ -183,12 +183,16 @@ typedef void (*net_route_cb_t)(struct net_route_entry *entry,
-	struct net_if *iface;
+	/** Network interfaces for the route. */
+	union {
+		struct net_if *iface;
```
Why would we need the union, is `struct net_if *iface` used anywhere in the codebase after these changes? IMO we should avoid such ambiguities, it could be problematic in the maintenance. And it's all internal, no need to worry about backward compatibility I think.

konradderda replied:
Hmm... I thought of backward compatibility in case if an user already has a code that adds/removes mcast routes based on events/external notifications which in case of few routes with the same prefix would require to search for a given interface with `net_route_mcast_foreach()` as `net_route_mcast_lookup()` and `net_route_mcast_del()` find the first occurence of the prefix. However, I am completely open to removing this.

konradderda replied:
Union removed.

jukkar commented:
ping @tbursztyka you have been working with routing code recently, please take a look

pdgendt commented on subsys/net/ip/route.c:
```diff
@@ -729,6 +729,55 @@ int net_route_foreach(net_route_cb_t cb, void *user_data)
+		return true;
+	}
+
+	for (int i = 0; i < CONFIG_NET_MCAST_ROUTE_MAX_IFACES; ++i) {
```
Nit, there's a helper macro for these kind of arrays.

```c
	ARRAY_FOR_EACH(entry->ifaces, i) {
		//...
	}
```

konradderda replied:
I have applied the remark and also added new first commit that uses the ARRAY_FOR_EACH_PTR() macro for all multicast routing related functions

kderda committed:
net: ipv6: use ARRAY_FOR_EACH_PTR() for iteration over mcast routes

This commmit changes the way of iterating over multicast routing
entries from explicit for-loop to ARRAY_FOR_EACH_PTR().

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>
